From 9f3551b4f5cf5ef681bc697d4de3f2d4882aab6e Mon Sep 17 00:00:00 2001
From: Adam Tkac <vonsch@gmail.com>
Date: Fri, 21 Sep 2012 12:39:57 +0200
Subject: [PATCH] Bump NVR to 2.0

Signed-off-by: Adam Tkac <vonsch@gmail.com>
---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 6686310..cff5526 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,5 +1,5 @@
 AC_PREREQ([2.59])
-AC_INIT([bind-dyndb-ldap], [1.1.0rc1], [freeipa-devel@redhat.com])
+AC_INIT([bind-dyndb-ldap], [2.0], [freeipa-devel@redhat.com])
 
 AM_INIT_AUTOMAKE([-Wall foreign dist-bzip2])
 
-- 
1.7.11.4

