From 1b5105e3ab92f2a898313da5f7e20e6f3e9d1d2a Mon Sep 17 00:00:00 2001
From: Petr Spacek <pspacek@redhat.com>
Date: Mon, 24 Feb 2014 16:48:09 +0100
Subject: [PATCH] Drop unnecessary #define _BSD_SOURCE.

Signed-off-by: Petr Spacek <pspacek@redhat.com>
---
 src/krb5_helper.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/krb5_helper.c b/src/krb5_helper.c
index d1787209483f2ae49b480492290ff5d4bafc677c..71f4fff9fec551abbd81e25c59de80d2ded0dfc6 100644
--- a/src/krb5_helper.c
+++ b/src/krb5_helper.c
@@ -15,8 +15,6 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
  */
 
-#define _BSD_SOURCE
-
 #include <isc/util.h>
 #include <string.h>
 #include <stdlib.h>
-- 
1.8.5.3

